docs/proposals/kubelet-systemd.md

```diff
@@ -0,0 +1,181 @@
+
+* `/sys/fs/cgroup/important/`
+* `/sys/fs/cgroup/unimportant/`
+
+In a unified hierarchy, a process should only belong to a single node in the `cgroup` tree.
+
+The Kernel direction for `cgroup` management is to promote a single-writer model rather than allowing multiple processes to independently
+write to parts of the file-system.
+
+## cgroup management with systemd and container runtimes
+
+By default, `systemd` defines three slices:
+
+* `system.slice` - contains system and application daemons
+* `user.slice` - user login related processes
+* `machine.slice` - virtual machines or containers that have init systems (not application space containers)
+
+A `slice` corresponds to an inner-node in the `cgroup` hierarchy.
+
+For example, the `system.slice` is represented as `/sys/fs/cgroup/<controller>/system.slice` in the tree.
+
+A `slice` is nested in the hierarchy by it's naming convention.
+
+For example, the `system-foo.slice` is represented as `/sys/fs/cgroup/<controller>/system.slice/system-foo.slice/`.
+
+Finally, to ensure everyone is on the same page, if `docker.service` is associated with `system.slice` and the user launches a container via
+`docker run -m 100M busybox`, a `scope` will be created because the process was not launched by `systemd` itself in the `slice` associated with the launching daemon.
+
+For example:
+
+`/sys/fs/cgroup/system.slice/docker-40c5264f71043fe15193603a05c7bf2a6b7f42b06951a7b59357cc7c2a380f04.scope`
+
+## kubelet cgroup driver
+
+The `kubelet` manipulates the `cgroup` file-system directly during initial bootstrapping.
+
+This proposal proposes that the `kubelet` accept a new flag:
+
+* `--cgroup-driver="": driver to use for cgroup manipulation by the kubelet.  Possible values: cgroupfs, systemd`
+
+The `ContainerManager` interface in the `kubelet` should support an implementation for `cgroupfs` or `systemd`.
+
+Effort to share code across implementations should be made or separated out.
+
+For example, both implementations will most likely rely on a common set of kernel tunable flags.
+
+## kubelet cgroup bootstrapping (current behavior)
+
+During initial bootstrapping of the `kubelet`, it creates the following `cgroup` nodes:
+
+* `/kubelet`
+
+If the `docker` container runtime is used, it creates the following `cgroup` node, and attaches the `docker` pid.
+
+* `/docker-daemon` - a memory restriction is imposed that limits memory to max( machine capacity * .7, 150Mi), the OOM score is adjusted to -900
+
+If a `system-container` name is specified, it will create an additional `cgroup` node
+
+* `/<system-container-name>` - attaches all non-kernel host pids to facilitate accounting host daemon usage
+
+## kubelet cgroup bootstrapping under systemd
+
+If the `kubelet` is launched with the `--cgroup-driver=systemd`, the following behavior is proposed:
+
+* the `kubelet` detects the `.service` unit from which it was launched and uses that for monitoring its own resource usage
```

Comment:
Maybe specify behaviour if the kubelet and/or docker daemon were not launched from services.